spectator/registry.go

```diff
@@ -0,0 +1,214 @@
+type RegistryInterface interface {
+	GetLogger() Logger
+	SetLogger(logger Logger)
+	NewMeter(id *meter.Id, meterFactory MeterFactoryFun) Meter
+	NewId(name string, tags map[string]string) *meter.Id
+	Counter(name string, tags map[string]string) *meter.Counter
+	CounterWithId(id *meter.Id) *meter.Counter
+	Timer(name string, tags map[string]string) *meter.Timer
+	TimerWithId(id *meter.Id) *meter.Timer
+	Gauge(name string, tags map[string]string) *meter.Gauge
+	GaugeWithId(id *meter.Id) *meter.Gauge
+	MaxGauge(name string, tags map[string]string) *meter.MaxGauge
+	MaxGaugeWithId(id *meter.Id) *meter.MaxGauge
+	DistributionSummary(name string, tags map[string]string) *meter.DistributionSummary
+	DistributionSummaryWithId(id *meter.Id) *meter.DistributionSummary
+	PercentileDistributionSummary(name string, tags map[string]string) *meter.PercentileDistributionSummary
+	PercentileDistributionSummaryWithId(id *meter.Id) *meter.PercentileDistributionSummary
+	PercentileTimer(name string, tags map[string]string) *meter.PercentileTimer
+	PercentileTimerWithId(id *meter.Id) *meter.PercentileTimer
+}
+
+// MeterFactoryFun is a type to allow dependency injection of the function used to generate meters.
+type MeterFactoryFun func() Meter
+
+// Used to validate that Registry implements RegistryInterface at build time.
+var _ RegistryInterface = (*Registry)(nil)
+
+// Registry is the collection of meters being reported.
+type Registry struct {
+	config *Config
+	writer writer.Writer
+	quit   chan struct{}
+}
+
+// TODO The New* methods will remain as is. Only the Registry instantiated should be changed.
+
+// NewRegistryConfiguredBy loads a new Config JSON file from disk at the path specified.
+func NewRegistryConfiguredBy(filePath string) (*Registry, error) {
+	path := filepath.Clean(filePath)
+	/* #nosec G304 */
+	bytes, err := ioutil.ReadFile(path)
+	if err != nil {
+		return nil, err
+	}
+
+	var config Config
+	err = json.Unmarshal(bytes, &config)
+	if err != nil {
+		return nil, err
+	}
+
+	return NewRegistry(&config), nil
+}
+
+// NewRegistry generates a new registry from the config.
+//
+// If config.IsEnabled is unset, it defaults to an implementation that returns
+// true.
+//
+// If config.Log is unset, it defaults to using the default logger.
+func NewRegistry(config *Config) *Registry {
+	if config.Log == nil {
+		config.Log = defaultLogger()
+	}
+
+	printWriter := &writer.PrintWriter{}
+	r := &Registry{
+		config: config,
+		writer: printWriter,
+		quit:   make(chan struct{}),
+	}
+
+	return r
+}
+
+// GetLogger returns the internal logger.
+func (r *Registry) GetLogger() Logger {
+	return r.config.Log
+}
+
+func (r *Registry) SetLogger(logger Logger) {
+	r.config.Log = logger
+}
+
+// TODO investigate how to provide the same functionality for MeterFactory. Do we need to expose the writer?
+// NewMeter creates a new Meter using the provided meterFactory
+func (r *Registry) NewMeter(id *meter.Id, meterFactory MeterFactoryFun) Meter {
```

Comment:
I searched for usages of MeterFactoryFun and I only found one [here](https://sourcegraph.netflix.net/stash.corp.netflix.com/OCNAS/nflx-gnmi-gateway@da154f685cfb1d20233fa3e02b4e934eaf5cd022/-/blob/vendor/github.com/openconfig/gnmi-gateway/gateway/stats/persistent_gauge.go) (same implementation exists in other repos too).

I need to think more if we can remove it and provide a different alternative.

Reply:
It looks like part of that usage is creating a "persistent" gauge that does not reset to NaN after reading?

None of this is going to work like that though - this becomes stateless.

Agree with thinking on it some more, to make sure we understand what they are doing and why, but lean into removing and providing migration instructions for the conversion.